python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,539 @@
+        with self.slot_lock:
+            self.job_queue_count += len(testers)
+
+        # Create a local dictionary of tester names to job containers. Add this dictionary to a
+        # set. We will use this set as a way to gain access to their methods.
+        for tester in testers:
+            name_to_job_container[tester.getTestName()] = TesterData(tester, job_dag, self.options)
+            self.tester_datas.add(name_to_job_container[tester.getTestName()])
+
+        skipped_or_failed_testers = self.checkGroupFailures(name_to_job_container)
+
+        # Create a set of failing job containers
+        for failed_tester in skipped_or_failed_testers:
+            non_runnable_jobs.add(name_to_job_container[failed_tester.getTestName()])
+
+        # Create dependencies in the DAG shared between this group of testers
+        self.createDependencies(name_to_job_container)
+
+        # Iterate over the items in our non_runnable_jobs and handle any skipped dependencies
+        for job in non_runnable_jobs.copy():
+            additionally_skipped = self.processDownstreamTests(name_to_job_container[job.getTestName()])
+            non_runnable_jobs.update(additionally_skipped)
+            job_dag.delete_node_if_exists(name_to_job_container[job.getTestName()])
+
+        # Get a count of all the items still in the DAG. This will be the jobs that ultimately get queued
+        runnable_jobs = job_dag.size()
+
+        # Make sure we didn't drop a tester somehow
+        if len(non_runnable_jobs) + runnable_jobs != len(testers):
+            raise SchedulerError('Runnable tests in addition to Skipped tests does not match total scheduled test count!')
+
+        # Assign a status thread to begin work on any skipped/failed jobs
+        self.queueJobs(status_job_containers=non_runnable_jobs)
+
+        # Build our list of runnable jobs and set the tester's status to queued
+        job_list = []
+        if runnable_jobs:
+            job_list = job_dag.ind_nodes()
+            for job_container in job_list:
+                tester = job_container.getTester()
+                tester.setStatus('QUEUED', tester.bucket_pending)
+
+        # Queue runnable jobs
+        self.queueJobs(tester_job_containers=job_list)
+
+    def waitFinish(self):
+        """
+        Block while the job queue is not empty. Once empty, this method will begin closing down
+        the thread pools and perform a join. Once the last thread exits, we return from this
+        method.
+
+        There are two thread pools in play; the Tester pool which is performing all the tests,
+        and the Status pool which is handling the printing of tester statuses. Because the
+        Status pool will always have the last item needing to be 'printed', we close and join
+        the Tester pool first, and then we do the same to the Status pool.
+        """
+        while self.job_queue_count > 0:
```

Comment:
Can this ever get stuck? The answer might be no, but I'm just wondering if the `job_queue_count` is guaranteed to decrease over time.